### PR TITLE
[TFLite] Add extra operators in the importer.

### PR DESCRIPTION
**Pull request by mciprian13**
**Summary**
- Add logic for loading the following operators from TensorFlowLite models:
  - ResizeNearest
  - DepthToSpace
  - SpaceToDepth
  - Cast
  - Gather
  - GatherND
  - LogSoftmax
  - Select
  - StridedSlice (with limitations)
  - Shape
- For the `Gather` node I renamed the `batchDims` attribute to `axis` since that is the meaning of the attribute and that is what the implementation does according to both Caffe2, ONNX (https://github.com/onnx/onnx/blob/master/docs/Operators.md#gather) and TFLite (https://www.tensorflow.org/mlir/tfl_ops#tflgather_tflgatherop) operator definitions.
- For the `GatherND` node I fixed the Interpreter implementation and added more unit tests in OperatorTest.cpp (the implementation was not generally correct). Also added the CPU implementation which was missing.

**Test Plan**
Add unit tests in TFLiteImporterTest and OperatorTest.

### CODE REVIEW

**Comment by mciprian13** [comment]
@vuzelac-cadence FYI

**Comment by mciprian13** [comment]
@zrphercule Can you review this? Thanks!

**Comment by mciprian13** [comment]
@yinghai Could you review this please? Thanks!

File: lib/Graph/Graph.cpp [hunk 1]
```diff
@@ -2808,42 +2808,58 @@ CollectRpnProposalsNode *Function::createCollectRpnProposals(
...
+  std::vector<dim_t> outDims(outRank);
+  size_t outIdx = 0;
+  for (size_t idx = 0; idx < batchDims; ++idx) {
+    outDims[outIdx++] = dataDims[idx];
+  }
+  for (size_t idx = batchDims; idx < indicesDims.size() - 1; ++idx) {
+    outDims[outIdx++] = indicesDims[idx];
+  }
+  for (size_t idx = batchDims + indicesDimLast; idx < dataDims.size(); ++idx) {
+    outDims[outIdx++] = dataDims[idx];
```
Comment (yinghai) [inline lib/Graph/Graph.cpp:2859]: @khabinov please help review this part wrt semantics of GatherND. Rest part lgtm.

**Review by khabinov — COMMENTED** [review]
Overall, looks good to me! Left a couple of comments.

File: include/glow/Graph/Graph.h [hunk 2]
```diff
@@ -1616,18 +1616,27 @@ class Function final : public IRContainer {
...
-  /// batch, and the result is the concatenation of the operation on each sample
-  /// in the batch.
+  /// Given \p data tensor of rank r >= 1, and \p indices tensor of rank q,
+  /// gather entries of the \p axis dimension of data (default outer-most for
+  /// axis = 0) indexed by indices and concatenate them in the output tensor of
+  /// rank q + (r - 1).
+  /// https://github.com/onnx/onnx/blob/master/docs/Operators.md#gather
   GatherNode *createGather(llvm::StringRef name, NodeValue data,
-                           NodeValue indices, unsigned_t batchDims = 0);
+                           NodeValue indices, unsigned_t axis = 0);
```
[the remark this answers is not in the document]
Comment (mciprian13) [reply include/glow/Graph/Graph.h:1625]: Thanks for the confirmation.

File: lib/Importer/TFLiteModelLoader.cpp [hunk 3]
```diff
@@ -1978,27 +2017,142 @@ Error TFLiteModelLoader::loadSlice(const tflite::Operator *op,
...
+  if (alignCorners) {
+    LOG(WARNING) << opErrMsg(opInfo, "Option 'align_corners' is ignored!");
   }
 
-  bool half_pixel_centers = opts->half_pixel_centers();
-  if (half_pixel_centers) {
-    LOG(WARNING) << opErrMsg(opInfo, "Half Pixel Centers option is ignored!");
+  bool halfPixelCenters = opts->half_pixel_centers();
+  if (halfPixelCenters) {
+    LOG(WARNING) << opErrMsg(opInfo, "Option 'half_pixel_centers' is ignored!");
```
[the remark this answers is not in the document]
Comment (mciprian13) [reply lib/Importer/TFLiteModelLoader.cpp:2033]: The `half_pixel_centers` option is not supported by the Glow Resize nodes but when used it only slightly changes the result of the interpolation such that most probably the inference is not affected dramatically. I would prefer to have this as-is and let the user know about this limitation without issuing and error.

File: tests/unittests/OperatorTest.cpp [hunk 4]
```diff
@@ -5642,90 +5642,144 @@ TEST_P(OperatorTest, GatherDataInt8IdxInt64) {
...
+TEST_GATHER_ND(4, FloatTy, Int64ITy, float, int64_t)
+TEST_GATHER_ND(5, FloatTy, Int64ITy, float, int64_t)
+#endif
+
+TEST_GATHER_ND(1, Float16Ty, Int32ITy, float16_t, int32_t)
+TEST_GATHER_ND(2, Float16Ty, Int32ITy, float16_t, int32_t)
+TEST_GATHER_ND(3, Float16Ty, Int32ITy, float16_t, int32_t)
+TEST_GATHER_ND(4, Float16Ty, Int32ITy, float16_t, int32_t)
+TEST_GATHER_ND(5, Float16Ty, Int32ITy, float16_t, int32_t)
+
```
[the remark this answers is not in the document]
Comment (mciprian13) [reply tests/unittests/OperatorTest.cpp:5774]: I added the guard as above. In principle this shouldn't be needed if the indices involved in this particular test have small values but for the general case it is needed. Added it for consistency.

**Comment by mciprian13** [comment]
@yinghai @khabinov Thank you for reviewing this. If there are no other review items, can we have this landed then?

**Comment by mciprian13** [comment]
@khabinov Any problems with this PR? Can we have this merged? Thanks!

**Comment by khabinov** [comment]
Would it be possible to revert renaming for `batchDims` to `axis`? Personally I really like this renaming, however it would cause some internal problems that we won't be able to resolve at the moment. Reverting that renaming part of this PR should help.

**Comment by mciprian13** [comment]
@khabinov I am just curious what problems does the renaming cause. I bet is something related to old `batchDims` setter/getter which should be relatively easy to solve?

**Comment by khabinov** [comment]
Yeah, it is about backwards compatibility for some serialized model which is not that trivial to solve. This would require some investment from our side, however we don't have resources to work on it at the moment.

**Comment by mciprian13** [comment]
@khabinov If I don't rename it back but instead I add the old getter `getBatchDims()` (in addition to `getAxis()`) would this solve the problem? If not then what exactly is the problem? I prefer finding a workaround than renaming a parameter to a name which is confusing!

**Comment by mciprian13** [comment]
@khabinov I renamed back the member to `BatchDims` for the Gather node as you requested. Can you try again this PR? I would love to have this merged this week. Thanks!

**Comment by mciprian13** [comment]
@khabinov Could you please have another try with this PR? I think I have made all the requested changes. Thanks!

**Review by jfix71 — APPROVED** [review]

**Pull request closed without merging**